Implement the firebase api on UI layer

pokk opened the pull request:
This branch is still running, just avoid to beomce huge. The works as the following:

- Added the firebase api on UI layer.
    - Added retreive the playone list api.
    - Added retreive the joined playone list api.
    - Added retreive the favoriate playone list api.
- Added that separating the get the datasnap and transforming from a datasnap to an object.
- Added a typealias for the firebase callback.

pokk committed:
Implemented the firebase api.

pokk committed:
Merge branch 'develop/master' into develop/remote_api

pokk committed:
Modified the coding style and add the `inline` keyword.

pokk committed:
Modified the api to the strategy functions.

pokk committed:
Added the kotlin coroutine library into ui layer.

pokk committed:
Finished the strategies for snapshot to an object.

pokk committed:
Added a new abstract function.

pokk committed:
Added the favorite & join api for the playone list.

lekaha commented on mobile-ui/src/main/java/com/playone/mobile/ui/firebase/v1/PlayoneFirebaseV1.kt:
```diff
@@ -1,70 +1,192 @@
+     * @param errorCallback a function for getting a error when retrieving the data.
+     * @param block a really block to get a [List].
+     */
+    private fun userSnap2PlayoneList(
```
I this this is a valid naming according to https://android.github.io/kotlin-guides/style.html#naming-1
But I don't think in the name of `userSnap2PlayoneList` the `Snap2` is unusual constructs.
> Sometimes there is more than one reasonable way to convert an English phrase into camel case, such as when acronyms or unusual constructs like “IPv6” or “iOS” are present. To improve predictability, use the following scheme.

I know what you name this function but my opinion is using `userSnapToPlayoneList` is better.

lekaha commented on remote/src/main/java/com/playone/mobile/remote/PlayoneFirebase.kt:
```diff
@@ -8,4 +10,22 @@ abstract class PlayoneFirebase {
+    abstract fun getPlayoneList(
+        userId: Int,
+        callback: (model: List<PlayoneModel>) -> Unit,
+        errorCallback: (code: Int, msg: String, detail: String) -> Unit
```
How about use `FirebaseErrorCallback`?

pokk replied:
This error callback is for firebase, but I think it's too long. Might use typealias is the best choice.

lekaha commented on mobile-ui/src/main/java/com/playone/mobile/ui/firebase/v1/PlayoneFirebaseV1.kt:
```diff
@@ -1,70 +1,192 @@
+    //endregion
+
+    //region Strategies of snapshot to the object.
+    private fun snap2PlayoneList(dataSnapshot: DataSnapshot?) =
```
I this this is a valid naming according to https://android.github.io/kotlin-guides/style.html#naming-1
But I don't think in the name of userSnap2PlayoneList the Snap2 is unusual constructs.

Sometimes there is more than one reasonable way to convert an English phrase into camel case, such as when acronyms or unusual constructs like “IPv6” or “iOS” are present. To improve predictability, use the following scheme.

I know what you name this function but my opinion is using userSnapToPlayoneList is better.

lekaha commented on remote/src/main/java/com/playone/mobile/remote/PlayoneRemoteImpl.kt:
```diff
@@ -11,6 +11,7 @@ import com.playone.mobile.remote.model.PlayoneModel
 import com.playone.mobile.remote.model.UserModel
 import io.reactivex.Single
 
+@Suppress("NOTHING_TO_INLINE")
```
Can you let me know any reason must to use this suppress?

pokk replied:
If they're called by function, it costs more than inline function. I just reduce the writing for the developers.

lekaha replied:
I am not so sure what you mean.. 

> costs more than inline function

I just feel confused because the warning says there is `performance impact`. What do you mean cost?

> I just reduce the writing for the developers.

the usage of inline function and normal function are the same, right? What you mean `reduce`?

pokk replied:
- 1. The keyword `inline` is the same as C's `inline`. That meaning just a few codes, it could be assembled to the code and the performance is also better than call other address of the function pointer. If I misunderstand, please let me know it.

- 2. I prefer to use the way of Functional Programming, I'd like to create some convenient functions as like `toXXX`. That's easier to understand what's that meaning and to code, isn't it?

lekaha replied:
@pokk This is what I think, 
As we know `inline` function should be as small as possible, and its behavior is just as C's `inline` that is right. But in Kotlin, I think because there are some overhead of Kotlin lambda covert to Java bytecode(support 6, 7, 8)  so use `inline` we put the body of `lambda` into the calling place can avoid some overhead issues. That is what I think the warning reported and it is what the difference with `C` language. But for normal usage (without lambda parameters) and because the function body is very small It is better to just drop out that function, right?
Or if you want to reuse the code, I think you can make a normal function. 

However, I just think we should not easily put `Suppress` and ignore the warnings.

pokk replied:
@lekaha 

Okay, I totally agree with you this point. I could remove the annotation for the inline function warning.
I've not tested which one is slow or costly yet. I'll try to test it in the near future XD.

lekaha commented on remote/src/main/java/com/playone/mobile/remote/PlayoneServiceFirebaseImpl.kt:
```diff
@@ -11,7 +12,12 @@ import com.playone.mobile.remote.model.UserModel
+    override fun fetchPlayoneList(userId: Int) =
+        single<List<PlayoneModel>> { emitter ->
+            playoneFirebase.getPlayoneList(userId, emitter::onSuccess) { code, msg, detail ->
+                emitter.onError(Exception("error code: $code, msg: $msg, detail: $detail"))
```
I think we can declare our own Exception and passing what we need as parameters rather than compose a message string.

pokk replied:
That's a good idea. I'm not familiar to declare the Exception. If you have a good way, please let me know.

lekaha commented on mobile-ui/src/main/java/com/playone/mobile/ui/firebase/v1/PlayoneFirebaseV1.kt:
```diff
@@ -1,70 +1,192 @@
+    }
+    else {
+        userDataSnapshot(userId.toString(),
+                         {},  // This is redundant anonymous function for running strategy function.
```
Please follow this https://android.github.io/kotlin-guides/style.html#continuation-indent

lekaha commented on dependencies.gradle:
```diff
@@ -228,6 +229,8 @@ ext {
             firebaseDb:         "com.google.firebase:firebase-database:${firebaseVersion}",
             firebaseMsging:     "com.google.firebase:firebase-messaging:${firebaseVersion}",
+
+            coroutine:          "org.jetbrains.kotlinx:kotlinx-coroutines-core:${kotlinCoroutine}",
```
In this PR, have you already used this dependency?

pokk replied:
No, I've tried using it, but in the current case is inconvenient. I think near future I need it, just keep it there.

lekaha commented on mobile-ui/build.gradle:
```diff
@@ -96,6 +102,7 @@ dependencies {
     implementation mobileUiDependencies.javaxAnnotation
 
     implementation mobileUiDependencies.kotlin
+    implementation mobileUiDependencies.coroutine
```
In this PR, have you already used this dependency?

lekaha commented on remote/src/main/java/com/playone/mobile/remote/mapper/Extension.kt:
```diff
@@ -7,16 +9,16 @@ import com.playone.mobile.remote.model.NotificationPayloadModel
 import com.playone.mobile.remote.model.UserModel
 
-fun PlayoneEntity.toModel(mapper: PlayoneEntityMapper) = mapper.mapFromData(this)
+inline fun PlayoneEntity.toModel(mapper: PlayoneEntityMapper) = mapper.mapFromData(this)
```
Since there is no lambda function in the parameter, what you think? use `internal fun` instead of inline function? As you can see the warning, `inline work best for functions with lambda parameters`

pokk replied:
I know `inline function` is better for lambda parameters. But I don't expect `inline` works for `lambda parameters` here.

lekaha reviewed (changes requested):
Please check the comments

pokk committed:
Fixed https://github.com/lekaha/playone-android/pull/24#pullrequestreview-98446845.

- https://github.com/lekaha/playone-android/pull/24#discussion_r169853629
- https://github.com/lekaha/playone-android/pull/24#discussion_r169853684
- https://github.com/lekaha/playone-android/pull/24#discussion_r169854196
- https://github.com/lekaha/playone-android/pull/24#discussion_r169854336

pokk committed:
Removed the warning annotation. https://github.com/lekaha/playone-android/pull/24#discussion_r169853793
